Overlays for trigger objects

philiprodrigues opened the pull request:
This PR adds the trigger data objects (TriggerPrimitive, TriggerActivity and TriggerCandidate) to `detdataformats`, along with overlay classes to allow them to be saved into fragments in a manner similar to what's used for the electronics data formats. The implementation is a bit convoluted in order to make the objects convenient to access from DS code, while also providing the overlays in the format required. A detailed discussion of the implementation is given in `README.md` in the `include/detdataformats/trigger/` directory.

Some outstanding questions/TODOs:

* I've used a "flexible array member" in the overlay classes. This isn't standard C++, but is supported by GCC. I think this is the most "overlayish" way to have a variable-sized array of objects, but we could replace it with getters and setters acting on memory past the end of the overlay class proper if we want standard-compliant code.
* `TriggerCandidate`'s `detid` member was previously a `std::vector<detid_t>`, but serializing a vector member is tricky, so I just changed it to a plain `detid_t`. I think this will probably be fine, since we haven't yet been very clear about what information is contained in the `detid`, but this probably needs some thought.
* I copied the `Types.hpp` file over from `triggeralgs`, which contains some types copied from `daqdataformats`. It would be nice to not copy those. That would require either adding a dependency of `detdataformats` on `daqdataformats`, or moving this code to `daqdataformats` (or maybe something else).

alessandrothea commented:
@philiprodrigues Looks very nice. My only suggestion is to remove the `Overlay` postfix from the class names, as none of the other overlay classes in `detdataformats` have it.

philiprodrigues commented:
I made that change, and updated downstream packages (trigger and triggeralgs). I've also moved the version member to the start of TP/TA/TC, so that we can change the format in future if necessary, while allowing readers to know what version of the class they're reading. I followed the approach from `FragmentHeader` for that.

Please go ahead and merge if you're happy with the changes.
